### PR TITLE
remote: fix name and ID collisions of containers and pods

### PR DESCRIPTION
**Pull request by vrothberg**
Fix the look up of containers and pods in the remote client.  User input
can refer to both, names or IDs of containers and pods, so there is a fair
chance of collisions (e.g., "c1" name with a "c1...." ID).

Those collisions are well handled (and battle tested) in the local
client which is directly using the libpod backend.  Hence, the remote
client should not attempt to introduce its own logic to prevent bugs and
divergence between the local and the remote clients.  To prevent
collisions such as in #7837, do a container/pod inspect on the
user-provided input to find the corresponding ID and eventually do full
ID comparisons to avoid potential collisions with names.

Note that this has a cost that I am not entirely happy with.  Looking at
issue #7837, the collisions are happening when removing the two
containers.  Remote container removal is now very chatty with the server
as it first queries for all containers, then iterates over the provided
names or IDs and does a remote inspect to figure out the IDs and find a
matching container object.  However, remote removal could just pass the
names and IDs directly to the batch removal endpoint.  Querying for all
containers could be prevented if the batch removal endpoint would remove
all if the slice is empty.

In other words, the bug is fixed but there's room for performance
improvements.

Fixes: #7837
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@containers/podman-maintainers PTAL ... something to look for at other places as well  (e.g., volumes, networks, etc.).

File: pkg/domain/infra/tunnel/helpers.go [hunk 1]
```diff
@@ -2,74 +2,114 @@ package tunnel
...
+				filtered = append(filtered, pod)
 				found = true
 				break
 			}
+
 		}
+
 		if !found {
-			return nil, errors.Wrapf(define.ErrNoSuchPod, "unable to find pod %q", nameOrID)
+			return nil, errors.Wrapf(define.ErrNoSuchCtr, "unable to find pod %q", nameOrID)
```
Comment (vrothberg) [inline pkg/domain/infra/tunnel/helpers.go:111]: Copy error ... should be Pod :)

**Comment by vrothberg** [comment]
Maybe, container removal needs a similar batch-removal endpoint as we have for images?

**Comment by mheon** [comment]
I don't think fixing this by adding a batch endpoint is a good idea. This
code is used in dozens of places and we'd need new endpoints for each. I
think we may want to investigate improving the list endpoint so we only
need to make one call - adding a nameOrId filter maybe?

**Comment by mheon** [comment]
This change LGTM though. We can optimize later.

**Comment by vrothberg** [comment]
> I don't think fixing this by adding a batch endpoint is a good idea. This code is used in dozens of places and we'd need new endpoints for each. I think we may want to investigate improving the list endpoint so we only need to make one call - adding a nameOrId filter maybe?

Sounds good to me. I had the same idea  to extend the list endpoint (see code comments).

**Comment by vrothberg** [comment]
Failures look legit, will have a look

**Comment by vrothberg** [comment]
Updated, pod tests should pass now. Had to add some `errorhandlig.Containes` checks.

**Comment by vrothberg** [comment]
@baude @jwhonce @rhatdan PTAL
